### ISSUE TITLE
RibbonAutoConfiguration error at build time

### ISSUE THREAD

**Issue by VuDo98** [spring-projects-experimental/spring-graalvm-native#392]
Hi, I am getting this error at build time when using spring-cloud-starter-netflix-eureka-client with HOXTON.SR8. It looks like the error is caused by the dep itself. Is there any possible fix without waiting for the dep to be update or fixed?

`
>Processing type org.springframework.cloud.netflix.ribbon.RibbonAutoConfiguration
Analyzing org.springframework.cloud.netflix.ribbon.RibbonAutoConfiguration reached by [[Ctx:org.springframework.cloud.netflix.ribbon.RibbonAutoConfiguration-FromSpringFactoriesKey]]
found 4 hints on org.springframework.cloud.netflix.ribbon.RibbonAutoConfiguration:
>1) Hint{[osca.Conditional],inferred=[oscnr.RibbonAutoConfiguration$RibbonClassesConditions:ACS(CLS CONS)]}
>2) Hint{[oscnr.RibbonClients,osca.Import],inferred=[oscnr.RibbonClientConfigurationRegistrar:ACS(RES CLS CONS)]}
>3) Hint{[osbcp.EnableConfigurationProperties],specific=[osbcp.ConstructorBinding:AccessDesc:ACS(CLS CONS DMETHS FLDS)],inferred=[oscnr.ServerIntrospectorProperties:ACS(CLS CONS DMETHS),oscnr.RibbonEagerLoad
Properties:ACS(CLS CONS DMETHS)]}
>4) Hint{[osbcp.EnableConfigurationProperties,osca.Import],inferred=[osbcp.EnableConfigurationPropertiesRegistrar:ACS(RES CLS CONS)]}
>processing hint Hint{[osca.Conditional],inferred=[oscnr.RibbonAutoConfiguration$RibbonClassesConditions:ACS(CLS CONS)]}
attempting registration of 1 inferred types
attempting registration of 2 elements of annotation hint chain
In agent mode skipping org.springframework.cloud.netflix.ribbon.RibbonAutoConfiguration because in already existing configuration
processing hint Hint{[oscnr.RibbonClients,osca.Import],inferred=[oscnr.RibbonClientConfigurationRegistrar:ACS(RES CLS CONS)]}
attempting registration of 1 inferred types
will follow Type:org/springframework/cloud/netflix/ribbon/RibbonClientConfigurationRegistrar
attempting registration of 3 elements of annotation hint chain
In agent mode skipping org.springframework.cloud.netflix.ribbon.RibbonAutoConfiguration because in already existing configuration
processing hint Hint{[osbcp.EnableConfigurationProperties],specific=[osbcp.ConstructorBinding:AccessDesc:ACS(CLS CONS DMETHS FLDS)],inferred=[oscnr.ServerIntrospectorProperties:ACS(CLS CONS DMETHS),oscnr.Ri
bbonEagerLoadProperties:ACS(CLS CONS DMETHS)]}
attempting registration of 1 specific types
attempting registration of 2 inferred types
attempting registration of 2 elements of annotation hint chain
In agent mode skipping org.springframework.cloud.netflix.ribbon.RibbonAutoConfiguration because in already existing configuration
processing hint Hint{[osbcp.EnableConfigurationProperties,osca.Import],inferred=[osbcp.EnableConfigurationPropertiesRegistrar:ACS(RES CLS CONS)]}
attempting registration of 1 inferred types
will follow Type:org/springframework/boot/context/properties/EnableConfigurationPropertiesRegistrar
attempting registration of 3 elements of annotation hint chain
In agent mode skipping org.springframework.cloud.netflix.ribbon.RibbonAutoConfiguration because in already existing configuration`
>processHierarchy oscnr.RibbonAutoConfiguration
Fatal error:java.lang.IllegalStateException: java.lang.IllegalStateException: ERROR: in 'org.springframework.cloud.netflix.ribbon.RibbonAutoConfiguration' these methods are directly invoking methods marked
@Bean: [loadBalancerClient, ribbonApplicationContextInitializer] - due to the enforced proxyBeanMethods=false for components in a native-image, please consider refactoring to use instance injection. If you
are confident this is not going to affect your application, you may turn this check off using -Dspring.native.verify=false.
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
        at java.base/java.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:600)
        at java.base/java.util.concurrent.ForkJoinTask.get(ForkJoinTask.java:1006)
        at com.oracle.svm.hosted.NativeImageGenerator.run(NativeImageGenerator.java:483)
        at com.oracle.svm.hosted.NativeImageGeneratorRunner.buildImage(NativeImageGeneratorRunner.java:350)
        at com.oracle.svm.hosted.NativeImageGeneratorRunner.build(NativeImageGeneratorRunner.java:509)
        at com.oracle.svm.hosted.NativeImageGeneratorRunner.main(NativeImageGeneratorRunner.java:115)
        at com.oracle.svm.hosted.NativeImageGeneratorRunner$JDK9Plus.main(NativeImageGeneratorRunner.java:541)
Caused by: java.lang.IllegalStateException: ERROR: in 'org.springframework.cloud.netflix.ribbon.RibbonAutoConfiguration' these methods are directly invoking methods marked @Bean: [loadBalancerClient, ribbon
ApplicationContextInitializer] - due to the enforced proxyBeanMethods=false for components in a native-image, please consider refactoring to use instance injection. If you are confident this is not going to
 affect your application, you may turn this check off using -Dspring.native.verify=false.
        at org.springframework.graalvm.type.Type.verifyComponent(Type.java:2273)
        at org.springframework.graalvm.support.ResourcesHandler.processType(ResourcesHandler.java:1282)
        at org.springframework.graalvm.support.ResourcesHandler.processType(ResourcesHandler.java:960)
        at org.springframework.graalvm.support.ResourcesHandler.checkAndRegisterConfigurationType(ResourcesHandler.java:950)
        at org.springframework.graalvm.support.ResourcesHandler.processSpringFactory(ResourcesHandler.java:849)
        at org.springframework.graalvm.support.ResourcesHandler.processSpringFactories(ResourcesHandler.java:714)
        at org.springframework.graalvm.support.ResourcesHandler.register(ResourcesHandler.java:130)
        at org.springframework.graalvm.support.SpringFeature.beforeAnalysis(SpringFeature.java:107)
        at com.oracle.svm.hosted.NativeImageGenerator.lambda$runPointsToAnalysis$7(NativeImageGenerator.java:696)
        at com.oracle.svm.hosted.FeatureHandler.forEachFeature(FeatureHandler.java:70)
        at com.oracle.svm.hosted.NativeImageGenerator.runPointsToAnalysis(NativeImageGenerator.java:696)
        at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:558)
        at com.oracle.svm.hosted.NativeImageGenerator.lambda$run$0(NativeImageGenerator.java:471)
        at java.base/java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(ForkJoinTask.java:1407)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
        at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
`

**Comment by aclement** [comment]
Per the message you can run with `-Dspring.native.verify=false` to turn off the check, it may work. But yes, you are right, it highlights an issue in the dependency itself.

**Comment by VuDo98** [comment]
I did try to build with `-Dspring.native.verify=false` but it still give the same error as if it is still verifying. Please correct me if my understanding is wrong.

**Comment by gonella** [comment]
Same here. I used -Dspring.native.verify=false but it seems getting ignored. Besides that, what is exactly option to solve this issue? Is there any wiki showing guidelines to convert it?

Should I convert @Bean to what?  When mentioned about "instance injection", we need to add bean programmatically to Spring Web App Context? 

with something like:

AutowireCapableBeanFactory.initializeBean(Object existingbean, String beanName)

**Comment by aclement** [comment]
```
Should I convert @bean to what? When mentioned about "instance injection", we need to add bean programmatically to Spring Web App Context?
```

That would be something like:

If you have this code below, which would trigger the verification problem (because the call to bar() is made directly, it normally requires proxy interception):

```
@Bean
public Foo foo() {
  return new Foo(bar());
}

@Bean
public Bar bar() {
  return new Bar();
}
```

switch it to this where `bar()` is not called explicitly.

```
@Bean
public Foo foo(Bar b) {
  return new Foo(b);
}

public Bar bar() {
  return new Bar();
}
```

Now the error here is telling us about it occurring in a separate library on your class path, so nothing you can do in your app, the library needs changes to work in this context.

Now I have found a route to the verifier that wasn't guarded, it should be guarded now so you can turn off the verifier in this scenario (if you use 0.9.0-SNAPSHOTs of spring-graalvm-native).

**Comment by sdeleuze** [comment]
Ok so `RibbonAutoConfiguration` was indeed using cross `@Bean` invocations which is not a construct supported on native, but Ribbon is not supported anymore in upcoming 2020.0 release train (to be release shortly) so I would advise to wait this release whihc brings improved native compatibility and use Spring Cloud Loadbalancer instead of Ribbon. If Spring Cloud Loadbalancer shipped with 2020.0 does not work please raise an issue.

**Comment by dsyer** [comment]
Spring Cloud Loadbalancer is also available in Hoxton (no guarantees it works with native images out of the box, but it's worth a try, and that would be a different issue anyway).
